matchcode/models.py

```diff
@@ -307,22 +315,90 @@ def match(cls, fingerprint, exact_match=False):
 
         # Step 3: order matches from lowest Hamming distance to highest Hamming distance
         # TODO: consider limiting matches for brevity
-        good_matches = cls.objects.none()
-        for hamming_distance, match in sorted(matches_by_hamming_distance.items()):
-            if hamming_distance == 0:
-                # If we have an exact match, return and disregard others
-                good_matches |= match
-                break
+        hamming_distances_and_matches = []
+        for hamming_distance, matches in sorted(matches_by_hamming_distance.items()):
+            hamming_distances_and_matches.append(
+                (hamming_distance, matches)
+            )
+
+        if TRACE:
+            for hamming_distance, matches in hamming_distances_and_matches:
+                for match in matches:
+                    dct = model_to_dict(match)
+                    logger_debug(
+                        cls.__name__,
+                        'match:',
+                        'step_3_hamming_distance:',
+                        hamming_distance,
+                        'step_3_matched_package:',
+                        dct
+                    )
+
+        # Step 4: use file heuristics to rank matches from step 3
+
+        # If we are not given resource data, return the matches we have
+        if not (resource and hamming_distances_and_matches):
+            if hamming_distances_and_matches:
+                pks = []
+                for _, matches in hamming_distances_and_matches:
+                    for match in matches:
+                        pks.append(match.pk)
+                return cls.objects.filter(pk__in=pks)
             else:
-                # If we don't have an exact match, add all close matches we have
-                good_matches |= match
+                return cls.objects.none()
+
+        resource_size = resource.size
+        matches_by_rank_attributes = defaultdict(list)
+        for hamming_distance, matches in hamming_distances_and_matches:
+            for match in matches:
+                matched_resource = match.package.resources.get(path=match.path)
+
+                if TRACE:
+                    logger_debug(
+                        cls.__name__,
+                        'match:',
+                        'step_4_matched_resource:',
+                        matched_resource
+                    )
+
+                # Compute size and name distance (similarity)
+                if matched_resource.is_file:
+                    size_distance = 1 - (resource_size / matched_resource.size)
```

Comment:
@pombredanne Should this be `1 - (resource_size / matched_resource.size)` or `1 - (matched_resource.size / resource_size)`?